Userland/Libraries/LibWeb/Infra/Numbers.cpp

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2023, Jonatan Klemets <jonatan.r.klemets@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/GenericLexer.h>
+#include <LibWeb/Infra/CharacterTypes.h>
+#include <LibWeb/Infra/Numbers.h>
+
+namespace Web::Infra {
+
+// This implements both the signed and unsigned parsing.
+// https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#rules-for-parsing-integers
+// https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#rules-for-parsing-non-negative-integers
+template<typename T>
+Optional<T> to_integer(StringView string)
+{
+    // 1. Let input be the string being parsed.
+    // 2. Let position be a pointer into input, initially pointing at the start of the string.
+    GenericLexer lexer { string };
+
+    // 4. Skip ASCII whitespace within input given position.
+    lexer.ignore_while(is_ascii_whitespace);
+
+    // 5. If position is past the end of input, return an error.
+    if (lexer.is_eof()) {
+        return {};
+    }
+
+    size_t start_index = lexer.tell();
+
+    // We skip steps 3, 6, and 9:
+    //   3. (Reorderd) Let sign have the value "positive".
+    //   6. If the character indicated by position (the first character) is a U+002D HYPHEN-MINUS character (-):
+    //   9. (Reorderd) If sign is "positive", return value, otherwise return the result of subtracting value from zero.
+    //
+    // Instead, if we parse a signed integer, then we include the sign character (if present) in the collect step
+    // (step 8) and lean on `AK::StringUtils::convert_to_int` to handle it for us.
+    //
+    // If we parse an unsigned integer we do more or less the exact same thing, but we don't include the sign and
+    // call `AK::StringUtils::convert_to_uint` instead.
```

Comment:
If possible, try not to rearrange the spec steps like this, it makes it more confusing to tell if the code is correct or not.

Instead, keep them in order, and add `NOTE:` comments to explain why each step is skipped or done differently. For example, something like:

```c++
// 3. Let sign have the value "positive".
// NOTE: Skipped, see comment on step 6.

// ...

// 6. If the character indicated by position (the first character) is a U+002D HYPHEN-MINUS character (-):
// NOTE: Instead, we include the sign character (if present) in the collect step
//       (step 8) and lean on `AK::StringUtils::convert_to_int` to handle it for us.
```